README.md

```diff
@@ -19,44 +19,80 @@ We assume that you already have:
+```rust
+use std::sync::Arc;
+use ark_ff::{FftField, PrimeField};
+use ark_poly::{EvaluationDomain, Radix2EvaluationDomain as D};
+use circuit_construction::*;
+use commitment_dlog::{commitment::CommitmentCurve, srs::SRS};
+use groupmap::GroupMap;
+use kimchi::verifier::verify;
+use mina_curves::pasta::{
+    fp::Fp,
+    vesta::{Affine, VestaParameters},
+};
 use oracle::{
-    constants::PlonkSpongeConstantsKimchi,
+    constants::*,
     sponge::{DefaultFqSponge, DefaultFrSponge},
 };
-use commitment_dlog::commitment::{b_poly_coefficients, ceil_log2, CommitmentCurve};
-
-type SpongeParams = PlonkSpongeConstantsKimchi;
-type BaseSponge = DefaultFqSponge<VestaParameters, SpongeParams>;
-type ScalarSponge = DefaultFrSponge<Fp, SpongeParams>;
-
-// compile the circuit
-let fp_sponge_params = oracle::pasta::fp_kimchi::params();
-let cs = ConstraintSystem::<Fp>::create(gates, vec![], fp_sponge_params, public_size).unwrap();
-
-// create an URS
-let mut urs = SRS::<Affine>::create(cs.domain.d1.size as usize);
-srs.add_lagrange_basis(cs.domain.d1);
-
-// obtain a prover index
-let prover_index = {
-    let fq_sponge_params = oracle::pasta::fq_kimchi::params();
-    let (endo_q, _endo_r) = endos::<Other>();
-    Index::<Affine>::create(cs, fq_sponge_params, endo_q, srs)
+type SpongeQ = DefaultFqSponge<VestaParameters, PlonkSpongeConstantsKimchi>;
+type SpongeR = DefaultFrSponge<Fp, PlonkSpongeConstantsKimchi>;
+
+pub fn circuit<
+    F: PrimeField + FftField,
+    Sys: Cs<F>,
+>(
+    witness: Option<F>,
+    sys: &mut Sys,
+    public_input: Vec<Var<F>>,
+) {
+    let three = sys.constant(3u32.into());
+    let first_public_input = public_input[0];
+    let witness = sys.var(|| witness.unwrap());
+
+    //TODO replace the following free variable once the "add" gate constraint is ready to use.
+    let result = sys.var(|| {
+        first_public_input.val() + witness.val()
+    });
+    sys.assert_eq(result, three);
+}
+
+// create SRS
+let srs = {
+    //TODO how to determine depth value? it throws error when the depth is too large
+    let mut srs = SRS::<Affine>::create(1 << 3); // 2^3 = 8
+    srs.add_lagrange_basis(D::new(srs.g.len()).unwrap());
+    Arc::new(srs)
 };
-
-// obtain a verifier index
-let verifier_index = prover_index.verifier_index();
-
-// create a proof
+let public_inputs = vec![1i32.into()];
 let group_map = <Affine as CommitmentCurve>::Map::setup();
-let proof =  ProverProof::create::<BaseSponge, ScalarSponge>(
-    &group_map, witness, &prover_index);
 
-// verify a proof
-verify::<Affine, BaseSponge, ScalarSponge>(&group_map, verifier_index, proof).unwrap();
+// generate circuit and index
+let prover_index = generate_prover_index::<FpInner, _>(
+    //TODO do these arguments have to be provided?
+    srs,
+    &fp_constants(),
+    &oracle::pasta::fq_kimchi::params(),
+    //TODO should this be encapsulated?
+    public_inputs.len(),
+    |sys, p| circuit::<_, _>(None, sys, p),
+);
```

Comment:
In the future it’ll be nice if we have a builder pattern for that as well

Reply:
And to not have the user have to figure out all of these params

Reply:
Yes, I think a builder pattern makes sense.